lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp

```diff
@@ -119,118 +131,579 @@ class TensorEmptyConversionPattern
 
 // TTIR::ToLayoutOp is a rather generic op that dictates how all the layout
 // properties of a tensor should be set. However, in TTNN world, multiple APIs
-// are required to achieve an arbitrary layout. There are two main distinct
-// paths in this conversion pattern:
-//
-// 1. If the layout calls for device memory, we will call TTNN::ToLayoutOp and
-//    TTNN::ToDeviceOp to achieve the desired layout.
-//
-// 2. If the layout calls for system memory, we will call TTNN::ToLayoutOp to
-//    change the tensor to RowMajor layout, and then the TTNN::FromDeviceOp to
-//    move to host memory
-//
+// are required to achieve an arbitrary layout, namely:
+// - ToLayoutOp: to set the layout (ROW_MAJOR, TILE) of the tensor
+// - TypecastOp: to change the data type of the tensor
+// - ToDeviceOp: to move the tensor to a specific device
+// - FromDeviceOp: to move the tensor from a specific device to host
+// - ToMemoryConfigOp: to set the memory configuration (dram, l1, interleaved,
+// sharded) of the tensor
```

Comment:
Let's zoom about this, I'd like to understand what the intent here is.

Reply:
Sure I'd be happy to zoom. In the meantime hopefully this [comment](https://github.com/tenstorrent/tt-mlir/pull/840#issuecomment-2387283194) clarifies things a bit.

Reply:
I don't want us to granularize this way for reasons mentioned in one of the comments above - we should model the `from_torch` approach, which is what the current code tries to do. Additionally, this piece of the change introduces a lot more code, which I find hard to read and maintain.

Can you post a list of specific scenarios we want to support, but are not currently supported?